server/app/services/program/predicate/PredicateGenerator.java

```
@@ -0,0 +1,290 @@
+              "Missing scalar or operator for predicate update form: %s", predicateForm.rawData()));
+    }
+
+    try {
+      return Pair.of(Scalar.valueOf(rawScalarValue), Operator.valueOf(rawOperatorValue));
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(
+          String.format(
+              "Bad scalar or operator for predicate update form: %s", predicateForm.rawData()));
+    }
+  }
+
+  private static PredicateDefinition.PredicateFormat detectFormat(
+      Multimap<Integer, LeafOperationExpressionNode> leafNodes) {
+    if (leafNodes.size() > 1
+        || leafNodes.keySet().stream()
+            .map(leafNodes::get)
+            .filter(
+                (Collection<LeafOperationExpressionNode> leafNodeGroup) -> leafNodeGroup.size() > 1)
+            .findAny()
+            .isPresent()) {
+      return PredicateDefinition.PredicateFormat.OR_OF_SINGLE_LAYER_ANDS;
+    }
+
+    return PredicateDefinition.PredicateFormat.SINGLE_QUESTION;
+  }
+
+  /**
+   * Parses the given value based on the given scalar type and operator. For example, if the scalar
+   * is of type LONG and the operator is of type ANY_OF, the value will be parsed as a list of
+   * comma-separated longs.
+   *
+   * <p>If value is the empty string, then parses the list of values instead.
+   */
+  // TODO: make this private once the old predicate logic is removed from
+  // AdminProgramBlockPredicatesController
+  public static PredicateValue parsePredicateValue(
+      Scalar scalar, Operator operator, String value, List<String> values) {
+
+    // If the scalar is SELECTION or SELECTIONS then this is a multi-option question predicate, and
+    // the right hand side values are in the `values` list rather than the `value` string.
+    if (scalar == Scalar.SELECTION || scalar == Scalar.SELECTIONS) {
+      ImmutableList.Builder<String> builder = ImmutableList.builder();
+      return PredicateValue.listOfStrings(builder.addAll(values).build());
+    }
```

if it's not SELECTION and there values then should throw an exception.

---

Prefer to avoid refactoring this code until the legacy code has been deleted

---

TODO maybe? fine if there isn't an issue number

---

done